txscript/pkscript.go

```diff
@@ -0,0 +1,245 @@
+		return false
+	}
+}
+
+// Class returns the script type.
+func (s PkScript) Class() ScriptClass {
+	return s.class
+}
+
+// Script returns the script as a byte slice without any padding.
+func (s PkScript) Script() []byte {
+	var script []byte
+
+	switch s.class {
+	case PubKeyHashTy:
+		script = make([]byte, pubKeyHashLen)
+		copy(script, s.script[:pubKeyHashLen])
+
+	case WitnessV0PubKeyHashTy:
+		script = make([]byte, witnessV0PubKeyHashLen)
+		copy(script, s.script[:witnessV0PubKeyHashLen])
+
+	case ScriptHashTy:
+		script = make([]byte, scriptHashLen)
+		copy(script, s.script[:scriptHashLen])
+
+	case WitnessV0ScriptHashTy:
+		script = make([]byte, witnessV0ScriptHashLen)
+		copy(script, s.script[:witnessV0ScriptHashLen])
+
+	default:
+		// Unsupported script type.
+		return nil
+	}
+
+	return script
+}
+
+// Address encodes the script into an address for the given chain.
+func (s PkScript) Address(chainParams *chaincfg.Params) (btcutil.Address, error) {
+	_, addrs, _, err := ExtractPkScriptAddrs(s.Script(), chainParams)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse address: %v", err)
+	}
+
+	return addrs[0], nil
+}
+
+// String returns a hex-encoded string representation of the script.
+func (s PkScript) String() string {
+	str, _ := DisasmString(s.Script())
+	return str
+}
+
+// ComputePkScript computes the pkScript of an transaction output by looking at
+// the transaction input's signature script or witness.
+//
+// NOTE: Only P2PKH, P2SH, P2WSH, and P2WPKH redeem scripts are supported.
+func ComputePkScript(sigScript []byte, witness wire.TxWitness) (PkScript, error) {
+	var pkScript PkScript
+
+	// Ensure that either an input's signature script or a witness was
+	// provided.
+	if len(sigScript) == 0 && len(witness) == 0 {
+		return pkScript, ErrUnsupportedScriptType
+	}
+
+	// We'll start by checking the input's signature script, if provided.
+	switch {
+	// If a signature script is provided with a length long enough to
+	// represent a P2PKH script, then we'll attempt to parse the compressed
+	// public key from it.
+	case len(sigScript) >= pubKeyHashSigScriptLen:
+		// The public key should be found as the last part of the
+		// signature script. We'll attempt to parse it to ensure this is
+		// a P2PKH redeem script.
+		pubKey := sigScript[len(sigScript)-compressedPubKeyLen:]
+		_, err := btcec.ParsePubKey(pubKey, btcec.S256())
+		if err == nil {
+			pubKeyHash := hash160(pubKey)
+			script, err := payToPubKeyHashScript(pubKeyHash)
+			if err != nil {
+				return pkScript, err
+			}
+
+			pkScript.class = PubKeyHashTy
+			copy(pkScript.script[:], script)
+			return pkScript, nil
+		}
+
+		// If it isn't, we'll assume it is the P2SH redeem script.
+		fallthrough
+
+	// If we failed to parse a compressed public key from the script in the
+	// case above, or if the script is smaller in length than that of a
+	// P2PKH one, we'll assume it's P2SH.
+	case len(sigScript) > 0:
+		scriptHash := hash160(sigScript)
```

Comment:
This is incorrect. Instead we need to take _only_ the first data push in the script and hash that. Otherwise, if this is vanilla p2sh, then we'll hash the redeem script _and_ any witness data.

Reply:
So we need to parse the data pushes in order to do this properly.

Reply:
Fixed.